### PR TITLE
Use a better data structure for SignedSubmissions instead of Vec

### PR DESCRIPTION
**Pull request by coriolinus**
Invent a new `SignedSubmissions` data structure which acts like a `BoundedBTreeSet` for the signed submissions, but has the property that it only reads a single submission at a time. This gives it the following properties:

- Take advantage of existing code to handle the bounding, instead of re-inventing the wheel.
- Take advantage of `O(n log n)` sorting properties for `BTreeMap` instead of the `O(n*n)` property we had for the vector.
- Substantially improve the worst-case memory use. We have to read more times, but each read is only a fraction as large.

### CODE REVIEW

**Comment by coriolinus** [comment]
> how was sorting done before?

Insertion-sort, essentially, using a highly idiosyncratic method.

https://github.com/paritytech/substrate/blob/88f856a2d05ac4b04b582d56207a3646ee90c8ea/frame/election-provider-multi-phase/src/signed.rs#L183-L199

An acknowledged consequence of this PR is that we'll end up using a total ordering by score instead. I don't think that this will really hurt us or change the incentives, though it's likely to change the actual winner of at least some elections.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 1]
```diff
@@ -43,12 +43,42 @@ pub struct SignedSubmission<AccountId, Balance: HasCompact, CompactSolution> {
...
+	CompactSolution: Ord,
+{
+	fn cmp(&self, other: &Self) -> Ordering {
+		self.solution
+			.score
+			.cmp(&other.solution.score)
+			.and_then(|| self.solution.cmp(&other.solution))
+			.and_then(|| self.deposit.cmp(&other.deposit))
+			.and_then(|| self.reward.cmp(&other.reward))
+			.and_then(|| self.who.cmp(&other.who))
```
[the remark this answers is not in the document]
Comment (coriolinus) [reply frame/election-provider-multi-phase/src/signed.rs:60]: Not really; we don't know the appropriate threshold at this point.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 2]
```diff
@@ -43,6 +49,34 @@ pub struct SignedSubmission<AccountId, Balance: HasCompact, CompactSolution> {
...
+	for SignedSubmission<AccountId, Balance, CompactSolution>
+where
+	AccountId: Ord,
+	Balance: Ord + HasCompact,
+	CompactSolution: Ord,
+{
+	fn cmp(&self, other: &Self) -> Ordering {
+		self.solution
+			.score
+			.cmp(&other.solution.score)
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/signed.rs:62]: this cmp is not taking `T::SolutionImprovementThreshold` into account.

Comment (coriolinus) [reply frame/election-provider-multi-phase/src/signed.rs:62]: That's true, because 
1. `SignedSubmission` is not defined in terms of `T`, and 
2. We can't really produce a stable total ordering if we _do_ take the `SolutionImprovementThreshold` into account.

Comment (kianenigma) [reply frame/election-provider-multi-phase/src/signed.rs:62]: That's informative, but not a solution :D 

1. You could make `SignedSubmission` have a `T: Get<Threshold>` and contain an additional `PhantomData<T>`
2. I can't see why. We have `ThresholdOrd` with the same API as `Ord`. An implementation of `Ord` can use `ThresholdOrd`


All in all, this is [uncool](https://www.youtube.com/watch?v=dy5S5kFZkzo); we can't throw away the entire use case of `SolutionImprovementThreshold` without having any second thought about it.

Comment (kianenigma) [reply frame/election-provider-multi-phase/src/signed.rs:62]: Sorry, I was wrong: I am reading the rest of the code, and I realized that we do take into account later when there is a congestion and one solution needs to be ejected. I think that's enough.

Comment (coriolinus) [reply frame/election-provider-multi-phase/src/signed.rs:62]: I didn't know about `ThresholdOrd`, which is cool, but it doesn't entirely solve the problem. According to the definition of `ThresholdOrd`, two items are equal if the greater is in the range of the lesser plus the threshold. `BTreeSet` is likely to get very confused if `Ord` says two items are equal but `Eq` disagrees. 

In any case, I'm not proposing to throw away `SolutionImprovementThreshold` entirely; there already exists some code checking it, and that could be extended to cover the full range of cases. It's just not in the `Ord` implementation.

https://github.com/paritytech/substrate/blob/33e6e456c55e4ae47c281ddb92e1a96e6ce85467/frame/election-provider-multi-phase/src/signed.rs#L217-L221

Comment (kianenigma) [reply frame/election-provider-multi-phase/src/signed.rs:62]: My comment was immature, I should have read the code; all good 💯

**Comment by kianenigma** [comment]
> As of [35277ea](https://github.com/paritytech/substrate/commit/35277ea6afc71e58a50a14bad127bc255f8314fc), only `all_in_one_signed_submission_scenario` still fails. I strongly suspect that this is related to the difference between `Ord` and the previously-used insertion sort for `SignedSubmission`, though I haven't yet proved it.
> 
> This is now the time to ask whether we really truly need to use that insertion sort, or whether we can accept that the top `T::SignedMaxSubmissions` are processed in true score order. It saves us some work if we can process them in the total ordering, but we can work around it if it's essential. In particular, would welcome @kianenigma 's thoughts.

It is fine to sort them in total order. 

The importance of  `ImprovementThreshold` is to prevent us from doing a full feasibility check for a solution that is a miniscule better. In the current scheme, we evaluate 1 signed solution anyhow, so total order is fine. 

Note that we still need to keen the `threshold_ord` for signed submissions competition.

**Comment by coriolinus** [comment]
Removing the polkadot PR label because this is merging into a different branch, which already takes care of the polkadot PR for us.

**Comment by coriolinus** [comment]
Sorry for the spam, all; accidentally merged in the wrong upstream and codeowners automatically tagged you.

File: frame/election-provider-multi-phase/src/lib.rs [hunk 3]
```diff
@@ -881,20 +885,22 @@ pub mod pallet {
...
+				(None, false) != (maybe_deposit, ejected_a_solution),
+				Error::<T>::SignedQueueFull,
+			);
+			signed_submissions.put();
+
+			if let Some(deposit_amount) = maybe_deposit {
+				// collect deposit. Thereafter, the function cannot fail.
+				T::Currency::reserve(&who, deposit_amount)
+					.map_err(|_| Error::<T>::SignedCannotPayDeposit)?;
+			}
```
Comment (coriolinus) [inline frame/election-provider-multi-phase/src/lib.rs]: Note that we haven't really done check-before-write here, because `signed_submissions` is three data structures in a trenchcoat pretending to be one. https://github.com/paritytech/substrate/blob/847b98cf5fa6f3a025b62ae9772eac27a0c531ef/frame/election-provider-multi-phase/src/signed.rs#L188-L191

If the deposit fails, then we end up with an incremented `SignedSubmissionsNextIndex` and an orphan entry in the `SignedSubmissionsMap`. The former is harmless. The latter occupies some chain storage space, but it will be cleared when the election is finalized.

Comment (coriolinus) [reply frame/election-provider-multi-phase/src/lib.rs]: Also: https://github.com/paritytech/substrate/pull/8933/commits/5b2a7c5f3fcb939f38249026dd1d8eb42de0f838 already moved the `signed_submissions.put()` to after the deposit.

Comment (coriolinus) [reply frame/election-provider-multi-phase/src/lib.rs]: 7f4a837ce fixes this.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 4]
```diff
@@ -51,8 +95,172 @@ pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
...
+		SignedSubmissions {
+			indices: SignedSubmissionIndices::<T>::get(),
+			next_idx: SignedSubmissionNextIndex::<T>::get(),
+			insertion_overlay: BTreeMap::new(),
+			deletion_overlay: BTreeSet::new(),
+		}
+	}
+
+	/// Put the signed submissions back into storage.
+	pub fn put(self) {
```
Comment (coriolinus) [inline frame/election-provider-multi-phase/src/signed.rs:127]: Ideally this would be automatic on `Drop`, but when I tried it, tests failed.

Comment (coriolinus) [reply frame/election-provider-multi-phase/src/signed.rs:127]: See https://github.com/paritytech/substrate/pull/8933/commits/3317a4bb4de2e77d5a7fff2154552a81ec081763.

**Comment by rphmeier** [comment]
> We have to read more times, but each read is only a fraction as large.

Aren't trie lookups significantly, as in hundreds of times, more expensive? What data do you have to indicate you are exploring this trade-off correctly?

**Comment by coriolinus** [comment]
@rphmeier The main impetus for this change comes from https://github.com/paritytech/substrate/pull/7910#issuecomment-853711559. In short: you're right that trie lookups dominate the runtime cost for small items. However, in this case, each `SignedSubmission` is very large, so decoding cost is a much more significant part of the work. Additionally, by decoding only one at a time instead of the full set of submissions, we can get a much better cap on total memory use.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 5]
```diff
@@ -51,8 +95,172 @@ pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
...
+		SignedSubmissionNextIndex::<T>::kill();
+		let insertion_overlay = sp_std::mem::take(&mut self.insertion_overlay);
+		SignedSubmissionsMap::<T>::drain()
+			.filter(move |(k, _v)| !self.deletion_overlay.contains(k))
+			.map(|(_k, v)| v)
+			.chain(insertion_overlay.into_iter().map(|(_k, v)| v))
+	}
+
+	/// Decode the length of the signed submissions without actually reading the entire struct into
+	/// memory.
```
Comment (emostov) [inline frame/election-provider-multi-phase/src/signed.rs:178]: Should this note that it does not include pending insertion or deletion overlay changes?

File: frame/support/src/storage/bounded_btree_map.rs [hunk 6]
```diff
@@ -407,4 +409,50 @@ pub mod test {
...
+
+		// but _can_ insert a distinct member which compares equal, though per the documentation,
+		// neither the set length nor the actual member are changed, but the value is
+		map.try_insert(Unequal(0, true), 6).unwrap();
+		assert_eq!(map.len(), 4);
+		let (zero_key, zero_value) = map.get_key_value(&Unequal(0, true)).unwrap();
+		assert_eq!(zero_key.0, 0);
+		assert_eq!(zero_key.1, false);
+		assert_eq!(*zero_value, 6);
+	}
```
Comment (emostov) [inline frame/support/src/storage/bounded_btree_map.rs:457]: Test is good. Not sure if its standard practice, but could be worth linking the docs for this non-identical key edge case: https://doc.rust-lang.org/std/collections/index.html#insert-and-complex-keys

File: frame/election-provider-multi-phase/src/signed.rs [hunk 7]
```diff
@@ -112,12 +315,13 @@ impl<T: Config> Pallet<T> {
 		// Any unprocessed solution is pointless to even consider. Feasible or malicious,
 		// they didn't end up being used. Unreserve the bonds.
 		let discarded = all_submissions.len();
-		for not_processed in all_submissions {
-			let SignedSubmission { who, deposit, .. } = not_processed;
+		for SignedSubmission { who, deposit, .. } in all_submissions.drain() {
```
Comment (emostov) [inline frame/election-provider-multi-phase/src/signed.rs:318]: When we drain with our btreemap (e.g. `SignedSubmissionsMap::<T>::drain()`), do we load/decode all the entries from trie into memory at once? Or does each iteration load/decode a single entry? (Maybe a better question is where could I find the `drain` impl? 😬)

Also, does the `SignedSubmission` get dropped from memory once we go to the next loop iteration?

**Comment by emostov** [comment]
With my level of understanding it looks good, but some Rust related details are still lost on me, so I am going to try and come back.

File: frame/support/src/storage/bounded_btree_map.rs (hunk 6 above)
Comment (coriolinus) [reply frame/support/src/storage/bounded_btree_map.rs:457]: `try_insert` does link to the docs of `BTreeMap::insert` and says that it has identical semantics. I'm not sure it's worth adding an additional, manual link.

File: frame/election-provider-multi-phase/src/signed.rs (hunk 7 above)
Comment (coriolinus) [reply frame/election-provider-multi-phase/src/signed.rs:318]: This calls the `fn drain` method defined on line 167. From there, it delegates to `SignedSubmissionMap::<T>::drain()`, which is an implementation of `IterableStorageMap::drain` which loads/decodes a single entry at a time.

And yes, each `SignedSubmission` gets dropped at the end of its loop iteration.

File: frame/election-provider-multi-phase/src/signed.rs (hunk 5 above)
Comment (coriolinus) [reply frame/election-provider-multi-phase/src/signed.rs:178]: https://github.com/paritytech/substrate/pull/8933/commits/58c96b44589945bb749a0892ec4fe5faa4084bbb

File: frame/election-provider-multi-phase/src/signed.rs [hunk 8]
```diff
@@ -43,6 +55,38 @@ pub struct SignedSubmission<AccountId, Balance: HasCompact, CompactSolution> {
 	pub solution: RawSolution<CompactSolution>,
 }
 
+impl<AccountId, Balance, CompactSolution> Ord
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/signed.rs:58]: would be better for readability if you put the impl after the thing is declared.

File: frame/election-provider-multi-phase/src/lib.rs [hunk 9]
```diff
@@ -1048,14 +1054,34 @@ pub mod pallet {
...
 	pub type SnapshotMetadata<T: Config> = StorageValue<_, SolutionOrSnapshotSize>;
 
-	/// Sorted (worse -> best) list of unchecked, signed solutions.
+	/// The next index to be assigned to an incoming signed submission.
+	///
+	/// We can't just use `SignedSubmissionIndices.len()`, because that's a bounded set; past its
+	/// capacity, it will simply saturate. We can't just iterate over `SignedSubmissionsMap`,
+	/// because iteration is slow. Instead, we store the value here.
+	#[pallet::storage]
+	pub(crate) type SignedSubmissionNextIndex<T: Config> = StorageValue<_, u32, ValueQuery>;
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:1063]: Is it correct to say this and the next one should never be accessed alone and only through `SignedSubmissions` struct?

Comment (kianenigma) [reply frame/election-provider-multi-phase/src/lib.rs:1063]: Looking at the logic of things like `map_take`, and the fact that you have internal overlays, it is probably the case.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 10]
```diff
@@ -51,8 +95,175 @@ pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
...
+			.map(|(_k, v)| v)
+			.chain(insertion_overlay.into_iter().map(|(_k, v)| v))
+	}
+
+	/// Decode the length of the signed submissions without actually reading the entire struct into
+	/// memory.
+	///
+	/// Note that if you hold an instance of `SignedSubmissions`, this function does _not_
+	/// track its current length. This only decodes what is currently stored in memory.
+	pub fn decode_len() -> Option<usize> {
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/signed.rs:182]: maybe even call it `decode_stored_len` or similar.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 11]
```diff
@@ -51,8 +95,175 @@ pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
...
+
+/// Mask type which pretends to be a set of `SignedSubmissionOf<T>`, while in fact delegating to the
+/// actual implementations in `SignedSubmissionIndices<T>`, `SignedSubmissionsMap<T>`, and
+/// `SignedSubmissionNextIndex<T>`.
+#[cfg_attr(feature = "std", derive(DebugNoBound))]
+pub struct SignedSubmissions<T: Config> {
+	indices: SubmissionIndicesOf<T>,
+	next_idx: u32,
+	insertion_overlay: BTreeMap<u32, SignedSubmissionOf<T>>,
+	deletion_overlay: BTreeSet<u32>,
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/signed.rs:112]: DQ: do the overlays need to be B-Trees themselves?

Comment (kianenigma) [reply frame/election-provider-multi-phase/src/signed.rs:112]: yeah, because we read from them as well in case of access.

File: frame/election-provider-multi-phase/src/signed.rs (hunk 4 above)
Comment (kianenigma) [reply frame/election-provider-multi-phase/src/signed.rs:127]: @apopiak has done similar work I belive. 

All in all, I am okay with not having this abstraction as well. `Drop` for storage writes sounds convenient, but I am a bit worried that it might have unforeseen negative consequences as well, since it is a bit of a paradigm shift.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 12]
```diff
@@ -51,8 +95,175 @@ pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
...
+			Ok(None) => {
+				// successfully inserted into the set; no need to take out weakest member
+				None
+			}
+			Err((score, insert_idx)) => {
+				// could not insert into the set because it is full.
+				// note that we short-circuit return here in case the iteration produces `None`.
+				// If there wasn't a weakest entry to remove, then there must be a capacity of 0,
+				// which means that we can't meaningfully proceed.
+				let (weakest_score, weakest_idx) = match self.indices.iter().next() {
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/signed.rs:216]: I haven't read the rest of this code, but it seems like you are making an invariant that the size of the queue is never 0? if so, I think this will be a nice feature to have. As-in, if the size of the queue is set to zero, then the governance is basically disabling signed solutions temporarily. 

If you have this invariant, please put it in `feasibility_chekc` as it is important.

Comment (kianenigma) [reply frame/election-provider-multi-phase/src/signed.rs:216]: Now reading the code, it seems like `return (false, None)` should handle everythign well.

File: frame/election-provider-multi-phase/src/signed.rs [hunk 13]
```diff
@@ -51,8 +95,175 @@ pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
...
+				}
+
+				self.indices.remove(&weakest_score);
+				self.indices
+					.try_insert(score, insert_idx)
+					.expect("just removed an item, we must be under capacity; qed");
+
+				// ensure that SignedSubmissionsMap never grows past capacity by taking out the
+				// weakest member here.
+				self.map_take(weakest_idx)
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/signed.rs:234]:
```suggestion
				let _removed = self.map_take(weakest_idx);
				debug_assert!(_removed.is_some());
```

File: frame/election-provider-multi-phase/src/signed.rs [hunk 14]
```diff
@@ -51,8 +95,175 @@ pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
...
+	/// Get the submission at a particular index.
+	fn map_get(&self, idx: u32) -> Option<SignedSubmissionOf<T>> {
+		self.insertion_overlay
+			.get(&idx)
+			.cloned()
+			.or_else(|| SignedSubmissionsMap::<T>::try_get(idx).ok())
+	}
+
+	/// Take the submission at a particular index.
+	fn map_take(&mut self, idx: u32) -> Option<SignedSubmissionOf<T>> {
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/signed.rs:147]: this names got me confused because I thought it is like `map_or_blah`, so there's a `map` operation, and then a `take`. But it seems like you meant "get from the map". I recommend maybe renaming it to something else, perhaps `get_solution` and `take_solution`

File: frame/election-provider-multi-phase/src/signed.rs [hunk 15]
```diff
@@ -51,8 +95,175 @@ pub type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
...
+		// we've taken out the weakest, so update the storage map and the next index
+		self.insertion_overlay.insert(self.next_idx, submission);
+		self.next_idx += 1;
+		(true, weakest)
+	}
+
+	/// Remove the signed submission with the highest score from the set.
+	pub fn pop_last(&mut self) -> Option<SignedSubmissionOf<T>> {
+		let (highest_score, idx) = self.indices.iter().rev().next()?;
+		let (highest_score, idx) = (*highest_score, *idx);
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/signed.rs:247]: wonder why https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.pop_last is not stable 🤔

File: frame/election-provider-multi-phase/src/signed.rs (hunk 7 above)
Comment (kianenigma) [reply frame/election-provider-multi-phase/src/signed.rs:321]: Good point, but this is safe since as mentioned we allocate one by one. 

> Maybe a better question is where could I find the drain impl


https://crates.parity.io/src/frame_support/storage/generator/map.rs.html#135-195

File: frame/election-provider-multi-phase/src/lib.rs [hunk 16]
```diff
@@ -881,20 +885,22 @@ pub mod pallet {
...
-				.get(index)
-				.map(|s| s.deposit)
-				.ok_or(Error::<T>::InvalidSubmissionIndex)?;
-			T::Currency::reserve(&who, deposit).map_err(|_| Error::<T>::SignedCannotPayDeposit)?;
-
-			// store the new signed submission.
-			<SignedSubmissions<T>>::put(signed_submissions);
+
+			let (maybe_deposit, ejected_a_solution) =
+				Self::insert_submission(&who, &mut signed_submissions, solution, size);
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:890]: I write like 5 different reviews here about how `insert_submission` is quite weird with regard to "check first, write last", as it clearly touches storage (like `unreserve`) but then there are errors after it. But then when you analyse each situation, it _seems_ to be correct and only touch the state when the `SignedQueueFull ` is not triggered. 

UPDATE: well, if the new solution is ejecting an old one, but it does not have enough deposit, then we unreserve the old one, but bail at reserving the new one. This is clearly a BIG issue, and would be good if you write a unit test for it as well (or make sure already unit tests catch it -- since they don't and that's also not a goog sign).

I recommend you to just throw my template of `insert_submission` with its return values out and rewrite this piece.

I think the simplest thing is to just dissolve `insert_submission` and paste it here, and make sure no write happens before all checks pass.  With your refactor, `insert_submission` is nicely slim.

File: frame/election-provider-multi-phase/src/lib.rs (hunk 9 above)
Comment (coriolinus) [reply frame/election-provider-multi-phase/src/lib.rs:1063]: Yes, that's absolutely the intent. That's also why they're as private as I can make them, and don't provide getters.

File: frame/election-provider-multi-phase/src/signed.rs (hunk 10 above)
Comment (coriolinus) [reply frame/election-provider-multi-phase/src/signed.rs:182]: I was basing it on [`StorageValue::decode_len`](https://substrate.dev/rustdocs/v3.0.0-monthly-2021-05/frame_support/storage/trait.StorageValue.html#method.decode_len).

File: frame/election-provider-multi-phase/src/signed.rs (hunk 11 above)
Comment (coriolinus) [reply frame/election-provider-multi-phase/src/signed.rs:112]: Yes, precisely.

File: frame/election-provider-multi-phase/src/signed.rs (hunk 13 above)
Comment (coriolinus) [reply frame/election-provider-multi-phase/src/signed.rs:234]: That's not a good idea; this line is part of the assignment to `weakest`, so not only is it not always going to be `Some(_)`, but we already use the value later.

Comment (kianenigma) [reply frame/election-provider-multi-phase/src/signed.rs:234]: yeah, oops.

File: frame/election-provider-multi-phase/src/signed.rs (hunk 14 above)
Comment (coriolinus) [reply frame/election-provider-multi-phase/src/signed.rs]: https://github.com/paritytech/substrate/pull/8933/commits/9388c2c037f8f6a9d435cd8b061b0e87aaf7d2e9

File: frame/election-provider-multi-phase/src/lib.rs (hunk 16 above)
Comment (coriolinus) [reply frame/election-provider-multi-phase/src/lib.rs]: Very good catch! Added test in https://github.com/paritytech/substrate/pull/8933/commits/a8e0441f922d55950f622cbcbc3fd114624fe24b; fixed in https://github.com/paritytech/substrate/pull/8933/commits/7dfc184e7b72452ee643f1a952719db140e1c0a8.

**Comment by coriolinus** [comment]
@kianenigma I think this is now ready for another look.

As for the `Drop` impl, I've realized that we don't want that after all. The whole point of the storage overlays is to ensure that we can return early without writing to storage. If writes were triggered on `Drop`, then we'd push the overlays to storage each time a `SignedSubmissions` exited scope, regardless of whether the exit was a success or an error. I haven't proved it (or even really tried to), but I suspect that the test failures given the `Drop` impl may be due to updating storage when we're not supposed to.

**Comment by kianenigma** [comment]
> but I suspect that the test failures given the Drop impl may be due to updating storage when we're not supposed to.

Yeah, this is the main potential footgun with `Drop` and I prefer keeping it explicit for now too.

File: frame/election-provider-multi-phase/src/lib.rs [hunk 17]
```diff
@@ -879,22 +883,37 @@ pub mod pallet {
...
-				.map(|s| s.deposit)
-				.ok_or(Error::<T>::InvalidSubmissionIndex)?;
-			T::Currency::reserve(&who, deposit).map_err(|_| Error::<T>::SignedCannotPayDeposit)?;
-
-			// store the new signed submission.
-			<SignedSubmissions<T>>::put(signed_submissions);
+			let (inserted, maybe_weakest) = signed_submissions.insert(submission);
+			let ejected_a_solution = maybe_weakest.is_some();
+
+			// it's an error if we neither inserted nor removed any submissions: this indicates
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:897]: WDYT? for me this is much easier to reason about, even if it is a few more lines of code: 

```
match (inserted, maybe_weakest)) => {
  (false, None) => { // Not inserted, none removed }
  (false, Some(weakest) => { unreachable!()??? }
  (true, None) => { // inserted and none was removed } 
  (true, Some(weakest) => { // inserted and someone was ejected }
}
```

which can be made into 

match (inserted, maybe_weakest)) => {
  (false, None) => { // Not inserted, none removed }
  (false, Some(weakest) => { unreachable!()??? }
  (true, maybe_removed) => { // inserted, maybe unreserve if maybe_removed.is_some() }
}
```

File: frame/election-provider-multi-phase/src/lib.rs (hunk 9 above)
Comment (kianenigma) [reply frame/election-provider-multi-phase/src/lib.rs:1076]: would be nice to maybe put it in the doc IN BOLD

File: frame/election-provider-multi-phase/src/signed.rs (hunk 10 above)
Comment (kianenigma) [reply frame/election-provider-multi-phase/src/signed.rs:182]: yeah, and I wanted to precisely _not_ be similar to that given that the semantic is slightly different :p

**Review by kianenigma — APPROVED** [review]
LGTM!

**Comment by kianenigma** [comment]
@coriolinus feel free to merge into the base branch when ready. Then you can proceed with urging the audit team to start their work on the main PR, and get the second approve.

I will re-visit the base PR one last time once this is merged into it.

File: frame/election-provider-multi-phase/src/lib.rs (hunk 17 above)
Comment (coriolinus) [reply frame/election-provider-multi-phase/src/lib.rs:897]: Doesn't that break down into what precisely exists now? We handle the `(false, None)` case with  an `ensure`, don't need to handle a `(false, Some(_))` unreachable case, and then handle the `(true, maybe_removed)` case inline. Agree that those are the cases we need to consider; I just think that the existing formulation is easier to read.

Comment (kianenigma) [reply frame/election-provider-multi-phase/src/lib.rs:897]: I read this in the downstream PR again and it was still confusing to me. Probably our definitions of "easy to read" differs. For me a readable code is one that can be easily understood with the least amount of time spent reading it. This entails being slightly verbose sometimes. 

A match statement's basis is that it cannot have any unhandled cases, and the reader can clearly see each case. The sequence of `if` statements took more time to reason about, at least in my case. 

I leave it up to you though, just wanted to share my detailed opinion.
